lsat/interceptor.go

```diff
@@ -0,0 +1,299 @@
+	"github.com/lightningnetwork/lnd/lnrpc/routerrpc"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/macaroons"
+	"github.com/lightningnetwork/lnd/zpay32"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// GRPCErrCode is the error code we receive from a gRPC call if the
+	// server expects a payment.
+	GRPCErrCode = codes.Internal
+
+	// GRPCErrMessage is the error message we receive from a gRPC call in
+	// conjunction with the GRPCErrCode to signal the client that a payment
+	// is required to access the service.
+	GRPCErrMessage = "payment required"
+
+	// AuthHeader is is the HTTP response header that contains the payment
+	// challenge.
+	AuthHeader = "WWW-Authenticate"
+
+	// MaxRoutingFee is the maximum routing fee in satoshis that we are
+	// going to pay to acquire an LSAT token.
+	// TODO(guggero): make this configurable
+	MaxRoutingFeeSats = 10
+
+	// PaymentTimeout is the maximum time we allow a payment to take before
+	// we stop waiting for it.
+	PaymentTimeout = 60 * time.Second
+
+	// manualRetryHint is the error text we return to tell the user how a
+	// token payment can be retried if the payment fails.
+	manualRetryHint = "consider removing pending token file if error " +
+		"persists. use 'listauth' command to find out token file name"
+)
+
+var (
+	// authHeaderRegex is the regular expression the payment challenge must
+	// match for us to be able to parse the macaroon and invoice.
+	authHeaderRegex = regexp.MustCompile(
+		"LSAT macaroon='(.*?)' invoice='(.*?)'",
+	)
+)
+
+// Interceptor is a gRPC client interceptor that can handle LSAT authentication
+// challenges with embedded payment requests. It uses a connection to lnd to
+// automatically pay for an authentication token.
+type Interceptor struct {
+	lnd         *lndclient.LndServices
+	store       Store
+	callTimeout time.Duration
+	lock        sync.Mutex
+}
+
+// NewInterceptor creates a new gRPC client interceptor that uses the provided
+// lnd connection to automatically acquire and pay for LSAT tokens, unless the
+// indicated store already contains a usable token.
+func NewInterceptor(lnd *lndclient.LndServices, store Store,
+	rpcCallTimeout time.Duration) *Interceptor {
+
+	return &Interceptor{
+		lnd:         lnd,
+		store:       store,
+		callTimeout: rpcCallTimeout,
+	}
+}
+
+// UnaryInterceptor is an interceptor method that can be used directly by gRPC
+// for unary calls. If the store contains a token, it is attached as credentials
+// to every call before patching it through. The response error is also
+// intercepted for every call. If there is an error returned and it is
+// indicating a payment challenge, a token is acquired and paid for
+// automatically. The original request is then repeated back to the server, now
+// with the new token attached.
+func (i *Interceptor) UnaryInterceptor(ctx context.Context, method string,
```

Comment:
We should add some context to the LSAT payment errors within this method. Some additional logging here would be nice too.

Reply:
An issue I ran into is that if we have a pending token but the call we're trying to access doesn't require one, then we'll still block the call from happening. Not sure how we can address this without introducing the whitelist in here as well.

Reply:
Good point, if no token token is required for a call then we shouldn't try to resume the token payment at that moment.
We can check it without knowing the whitelist, we just have to wait to run into a 402 first and only then resume the token payment.
I refactored the whole logic to make it (hopefully) easier to follow.